lightning/src/util/wakers.rs

```diff
@@ -0,0 +1,251 @@
+			// desired wait time has actually passed, and if not then restart the loop with a reduced wait
+			// time. Note that this logic can be highly simplified through the use of
+			// `Condvar::wait_while` and `Condvar::wait_timeout_while`, if and when our MSRV is raised to
+			// 1.42.0.
+			let elapsed = current_time.elapsed();
+			let result = guard.0;
+			if result || elapsed >= max_wait {
+				guard.0 = false;
+				return result;
+			}
+			match max_wait.checked_sub(elapsed) {
+				None => return result,
+				Some(_) => continue
+			}
+		}
+	}
+
+	/// Wake waiters, tracking that persistence needs to occur.
+	pub(crate) fn notify(&self) {
+		let mut persistence_lock = self.persist_pending.lock().unwrap();
+		persistence_lock.0 = true;
+		if let Some(future_state) = persistence_lock.1.take() {
+			future_state.lock().unwrap().complete();
+		}
+		mem::drop(persistence_lock);
+		self.persistence_condvar.notify_all();
+	}
+
+	/// Gets a [`Future`] that will get woken up with any waiters
+	pub(crate) fn get_future(&self) -> Future {
+		let mut persistence_lock = self.persist_pending.lock().unwrap();
+		if persistence_lock.0 {
+			Future {
+				state: Arc::new(Mutex::new(FutureState {
+					callbacks: Vec::new(),
+					complete: false,
+				}))
+			}
+		} else if let Some(existing_state) = &persistence_lock.1 {
+			Future { state: Arc::clone(&existing_state) }
+		} else {
+			let state = Arc::new(Mutex::new(FutureState {
+				callbacks: Vec::new(),
+				complete: false,
+			}));
+			persistence_lock.1 = Some(Arc::clone(&state));
+			Future { state }
+		}
+	}
+
+	#[cfg(any(test, feature = "_test_utils"))]
+	pub fn needs_persist(&self) -> bool {
+		self.persist_pending.lock().unwrap().0
+	}
+}
+
+/// A callback which is called when a [`Future`] completes.
+///
+/// Note that this MUST NOT call back into LDK directly, it must instead schedule actions to be
+/// taken later. Rust users should use the [`std::future::Future`] implementation for [`Future`]
+/// instead.
+///
+/// Note that the [`std::future::Future`] implementation may only work for runtimes which schedule
+/// futures when they receive a wake, rather than immediately executing them.
+pub trait FutureCallback : Send {
+	/// The method which is called.
+	fn call(&self);
+}
+
+pub(crate) struct FutureState {
+	callbacks: Vec<Box<dyn FutureCallback>>,
+	complete: bool,
+}
+
+impl FutureState {
+	fn complete(&mut self) {
+		for callback in self.callbacks.drain(..) {
+			callback.call();
+		}
+		self.complete = true;
+	}
+}
+
+/// A simple future which can complete once, and calls some callback(s) when it does so.
+pub struct Future {
+	state: Arc<Mutex<FutureState>>,
+}
+
+impl Future {
+	/// Registers a callback to be called upon completion of this future. If the future has already
+	/// completed, the callback will be called immediately.
+	pub fn register_callback(&self, callback: Box<dyn FutureCallback>) {
```

Comment:
Doesn't look like this is called. Is there any plans to use it in `BackgroundProcessor`?